client/v1/src/main/java/org/opennms/velocloud/client/v1/ApiExecutor.java

```diff
@@ -0,0 +1,120 @@
+import org.opennms.velocloud.client.v1.handler.ApiClient;
+
+import com.google.common.base.Objects;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+
+public class ApiExecutor {
+
+    /**
+     * Authentication parameter for ApiKeyAuth used in header parameter value
+     *
+     * @see org.opennms.velocloud.client.v1.handler.auth.ApiKeyAuth
+     */
+    public static final String AUTH_HEADER_PREFIX = "Token";
+    public static final String PATH = "portal/rest";
+
+    public static AllApi connectApi(final VelocloudApiClientCredentials credentials) {
+        final var client = new ApiClient();
+        client.setBasePath(URI.create(credentials.orchestratorUrl).resolve(PATH).toString());
+        client.setApiKeyPrefix(AUTH_HEADER_PREFIX);
+        client.setApiKey(credentials.apiKey);
+        return new AllApi(client);
+    }
+
+    private static class Key<K, V> {
+        final ApiCall<K, V> apiCall;
+        final VelocloudApiClientCredentials credentials;
+        final K parameter;
+
+        public Key(ApiCall<K, V> apiCall, VelocloudApiClientCredentials credentials, K parameter) {
+            this.apiCall = apiCall;
+            this.credentials = credentials;
+            this.parameter = parameter;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Key<?, ?> other = (Key<?, ?>) o;
+            return this.apiCall == other.apiCall
+                    && this.credentials.equals(other.credentials)
+                    && Objects.equal(parameter, other.parameter);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(apiCall, credentials, parameter);
+        }
+    }
+
+    final LoadingCache<Key<?, ?>, Object> cache;
+    final LoadingCache<VelocloudApiClientCredentials, AllApi> cacheApi;
+
+    public ApiExecutor(long expiringAfterMilliseconds) {
+        cacheApi = CacheBuilder.newBuilder()
+                .expireAfterWrite(expiringAfterMilliseconds, TimeUnit.MILLISECONDS)
+                .build(new CacheLoader<>() {
+                    @Override
+                    public AllApi load(VelocloudApiClientCredentials credentials) {
+                        return connectApi(credentials);
+                    }
+                });
+        cache = CacheBuilder.newBuilder()
+                .expireAfterWrite(expiringAfterMilliseconds, TimeUnit.MILLISECONDS)
+                .build(new CacheLoader<>() {
+                    @Override
+                    public Object load(Key<?, ?> key) throws Exception {
+                        final AllApi api = cacheApi.get(key.credentials);
+                        return key.apiCall.adapterCall(api, key.parameter);
+                    }
+                });
+    }
```

Comment:
I haven't used CacheLoaders before, so I want to make sure I understand what this is doing. When `cache.get()` is called for the first time, it calls the `load()` method here and saves the result for `expiringAfterMilliseconds` milliseconds. During that time, if `cache.get()` is called again with an equivalent key, it just returns that same result. So in effect, this will only call a given endpoint with the same credentials / parameters at most once every `expiringAfterMilliseconds` milliseconds.

Is that correct?

Reply:
> I haven't used CacheLoaders before, so I want to make sure I understand what this is doing. When `cache.get()` is called for the first time, it calls the `load()` method here and saves the result for `expiringAfterMilliseconds` milliseconds. During that time, if `cache.get()` is called again with an equivalent key, it just returns that same result. So in effect, this will only call a given endpoint with the same credentials / parameters at most once every `expiringAfterMilliseconds` milliseconds.
> 
> Is that correct?

Exactly. I also just add a test for the case of expired call